File: src/main/java/com/aurel/track/report/dashboard/TwoDimensionalStatistic.java
/**
 * Genji Scrum Tool and Issue Tracker
 * Copyright (C) 2015 Steinbeis GmbH & Co. KG Task Management Solutions

 * <a href="http://www.trackplus.com">Genji Scrum Tool</a>

 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/* $Id:$ */

package com.aurel.track.report.dashboard;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import com.aurel.track.Constants;
import com.aurel.track.admin.customize.category.filter.execute.loadItems.TooManyItemsToLoadException;
import com.aurel.track.beans.TPersonBean;
import com.aurel.track.fieldType.constants.SystemFields;
import com.aurel.track.json.JSONUtility;
import com.aurel.track.report.execute.ReportBean;
import com.aurel.track.util.IntegerStringBean;
import com.aurel.track.util.LabelValueBean;

/**
 * Can select filter, can select field for x-axis, field for y-axis,
 * sort-by, sort direction, show sums (yes, no), number of rows to display.
 * @author Adi
 *
 */
public class TwoDimensionalStatistic extends BasePluginDashboardView implements IssueProvider{
	public static final String SELECTED_FILTER = "selFilter";
	public static final String FIELDS_H = "fieldsH";
	public static final String FIELDS_V = "fieldsV";
	public static final String SELECTED_FIELD_H = "selFieldH";
	public static final String SELECTED_FIELD_V = "selFieldV";
	
	private static final String ALPHANUMERIC="alphanumeric";
	private static final String SORT_ORDER="sortOrder";
	private static final String ASC="asc";
	private static final String DESC="desc";
	
	protected static interface LINK_PARAMETERS {
		static String FIELD_VALUE_H= "fieldValueH";
		static String FIELD_VALUE_V= "fieldValueV";
		static String PROJECT = "project";
		static String RELEASE = "release";
	}

	protected boolean isUseConfig(){
		return true;
	}

	@Override
	protected String encodeJSONExtraData(Integer dashboardID,Map<String, Object> session, Map<String, String> parameters,
										 Integer projectID, Integer releaseID,Map<String,String> ajaxParams) throws TooManyItemsToLoadException{
		TPersonBean user = (TPersonBean) session.get(Constants.USER_KEY);
		StringBuilder sb=new StringBuilder();
		String filterId = parameters.get(SELECTED_FILTER);
		String filterTitle = "";
		JSONUtility.appendStringValue(sb,"filterId",filterId);
		if (filterId == null || "".equals(filterId)) {
			JSONUtility.appendStringValue(sb,"filterTitle",filterTitle);
		}else{
			Locale locale = user.getLocale();
			Integer fieldH=SystemFields.INTEGER_ISSUETYPE;
			Integer fieldV=SystemFields.INTEGER_STATE;
			String fieldHStr=parameters.get(SELECTED_FIELD_H);
			if(fieldHStr!=null){
				try{
					fieldH=Integer.valueOf(fieldHStr);
				}catch (Exception e) {
				}
			}
			String fieldVStr=parameters.get(SELECTED_FIELD_V);
			if(fieldVStr!=null){
				try{
					fieldV=Integer.valueOf(fieldVStr);
				}catch (Exception e) {
				}
			}

			String selectedSortByH=parameters.get("selectedSortByH");
			String selectedSortTypeH=parameters.get("selectedSortTypeH");
			boolean useSortOrderH=SORT_ORDER.equals(selectedSortByH);
			boolean sortDescH=DESC.equals(selectedSortTypeH);

			String selectedSortByV=parameters.get("selectedSortByV");
			String selectedSortTypeV=parameters.get("selectedSortTypeV");
			boolean useSortOrderV=SORT_ORDER.equals(selectedSortByV);
			boolean sortDescV=DESC.equals(selectedSortTypeV);


			Statistic2D statistic2d=TwoDimensionalStatisticBL.calculateStatistic(filterId,fieldH,sortDescH,useSortOrderH,fieldV,sortDescV,useSortOrderV, user, locale, projectID, releaseID);
			JSONUtility.appendJSONValue(sb,"statistic",TwoDimensionalStatisticBL.encodeJSON(statistic2d));
			filterTitle=BasePluginDashboardBL.getFilterTitle(filterId, (Locale)session.get(Constants.LOCALE_KEY));
			JSONUtility.appendStringValue(sb,"filterTitle", filterTitle);
		}

		return sb.toString();
	}

	@Override
	public String encodeJSONExtraDataConfig(Map<String,String> parameters, TPersonBean user, Integer entityId, Integer entityType){
		StringBuilder sb=new StringBuilder();
		String filterStr = parameters.get(SELECTED_FILTER);
		Integer filter=null;
		if(filterStr!=null){
			try{
				filter=Integer.parseInt(filterStr);
			}catch (Exception e){}
		}

		JSONUtility.appendIntegerValue(sb,SELECTED_FILTER, filter);
		String filterTitle=BasePluginDashboardBL.getFilterTitle(filterStr, user.getLocale());
		JSONUtility.appendStringValue(sb,"selFilterName", filterTitle);

		String selFieldHStr = parameters.get(SELECTED_FIELD_H);
		String selFieldVStr = parameters.get(SELECTED_FIELD_V);
		Integer selFieldH=null,selFieldV=null;
		if(selFieldHStr!=null){
			try{
				selFieldH=Integer.parseInt(selFieldHStr);
			}catch (Exception ex){}
		}
		if(selFieldVStr!=null){
			try{
				selFieldV=Integer.parseInt(selFieldVStr);
			}catch (Exception ex){}
		}
		if(selFieldH==null){
			selFieldH=SystemFields.ISSUETYPE;
		}
		if(selFieldV==null){
			selFieldV=SystemFields.STATE;
		}
		JSONUtility.appendIntegerValue(sb,SELECTED_FIELD_H,selFieldH);
		JSONUtility.appendIntegerValue(sb,SELECTED_FIELD_V,selFieldV);
		Locale locale=user.getLocale();

		List<IntegerStringBean> fieldsSelect = TwoDimensionalStatisticBL.getOptionFileds(locale);
		JSONUtility.appendIntegerStringBeanList(sb,FIELDS_H,fieldsSelect);
		JSONUtility.appendIntegerStringBeanList(sb,FIELDS_V,fieldsSelect);

		List<LabelValueBean> sortTypes=new ArrayList<LabelValueBean>();
		sortTypes.add(new LabelValueBean(localizeText("twoDimensionalStatistic.asc",locale),ASC));
		sortTypes.add(new LabelValueBean(localizeText("twoDimensionalStatistic.desc",locale),DESC));
		JSONUtility.appendLabelValueBeanList(sb,"sortTypes",sortTypes);
		String selectedSortTypeV=parameters.get("selectedSortTypeV");
		if(selectedSortTypeV==null||selectedSortTypeV.trim().length()==0){
			selectedSortTypeV=ASC;
		}
		JSONUtility.appendStringValue(sb,"selectedSortTypeV",selectedSortTypeV);
		String selectedSortTypeH=parameters.get("selectedSortTypeH");
		if(selectedSortTypeH==null||selectedSortTypeH.trim().length()==0){
			selectedSortTypeH=ASC;
		}
		JSONUtility.appendStringValue(sb,"selectedSortTypeH",selectedSortTypeH);


		List<LabelValueBean> sortBys=new ArrayList<LabelValueBean>();
		sortBys.add(new LabelValueBean(localizeText("twoDimensionalStatistic.alphanumeric",locale),ALPHANUMERIC));
		sortBys.add(new LabelValueBean(localizeText("twoDimensionalStatistic.sortOrder",locale),SORT_ORDER));
		JSONUtility.appendLabelValueBeanList(sb,"sortBys",sortBys);
		String selectedSortByV=parameters.get("selectedSortByV");
		if(selectedSortByV==null||selectedSortByV.trim().length()==0){
			selectedSortByV=ALPHANUMERIC;
		}
		JSONUtility.appendStringValue(sb,"selectedSortByV",selectedSortByV);
		String selectedSortByH=parameters.get("selectedSortByH");
		if(selectedSortByH==null||selectedSortByH.trim().length()==0){
			selectedSortByH=ALPHANUMERIC;
		}
		JSONUtility.appendStringValue(sb,"selectedSortByH",selectedSortByH);

		int prefWidth=700;
		int prefHeight=400;
		JSONUtility.appendIntegerValue(sb,"prefWidth",prefWidth);
		JSONUtility.appendIntegerValue(sb,"prefHeight",prefHeight);

		return sb.toString();
	}

	@Override
	public List<ReportBean> getIssues(Map<String, String> configParams, Map<String, String> filterParams, TPersonBean personBean, Locale locale) throws TooManyItemsToLoadException{
		return TwoDimensionalStatisticBL.getIssues(configParams, filterParams,personBean,locale);
	}
	public int[] getIssueNos(Map providerParams) {
		return null;
	}
	
	public static class Statistic2D{
		private String labelH;
		private String labelV;
		private List<LabelValueBean> optionsH;
		private List<LabelValueBean> optionsV;
		private List<Integer> totalH;
		private List<Integer> totalV;
		private LabelValueBean[][] values;
		public String getLabelH() {
			return labelH;
		}
		public void setLabelH(String labelH) {
			this.labelH = labelH;
		}
		public String getLabelV() {
			return labelV;
		}
		public void setLabelV(String labelV) {
			this.labelV = labelV;
		}
		public List<LabelValueBean> getOptionsH() {
			return optionsH;
		}
		public void setOptionsH(List<LabelValueBean> optionsH) {
			this.optionsH = optionsH;
		}
		public List<LabelValueBean> getOptionsV() {
			return optionsV;
		}
		public void setOptionsV(List<LabelValueBean> optionsV) {
			this.optionsV = optionsV;
		}
		public List<Integer> getTotalH() {
			return totalH;
		}
		public void setTotalH(List<Integer> totalH) {
			this.totalH = totalH;
		}
		public List<Integer> getTotalV() {
			return totalV;
		}
		public void setTotalV(List<Integer> totalV) {
			this.totalV = totalV;
		}
		public LabelValueBean[][] getValues() {
			return values;
		}
		public void setValues(LabelValueBean[][] values) {
			this.values = values;
		}
	}
}

File: src/main/webapp/js/ext/com/track/dashboard/TwoDimensionalStatistic.js
/**
 * Genji Scrum Tool and Issue Tracker
 * Copyright (C) 2015 Steinbeis GmbH & Co. KG Task Management Solutions

 * <a href="http://www.trackplus.com">Genji Scrum Tool</a>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/* $Id:$ */

Ext.define('js.ext.com.track.dashboard.TwoDimensionalStatistic',{
	extend:'js.ext.com.track.dashboard.DashboardRenderer',

	initComponent : function(){
		var me=this;
		me.items=me.createChildren();
		me.callParent();
		if(me.jsonData.filterTitle){
			me.title=me.title+"&nbsp;-&nbsp;<i>"+me.jsonData.filterTitle+"</i>";
		}
	},
	createChildren:function(){
		var me=this;
		if(me.jsonData.tooManyItems===true){
			return [me.createErrorCmp(getText('cockpit.err.tooManyItems'))];
		}
		var statistic=me.jsonData.statistic;
		if(statistic){
			return [me.createGrid()];
		}else{
			return [];
		}
	},
	doRefresh:function(jsonData){
		var me=this;
		me.jsonData=jsonData;
		me.removeAll();
		me.add(me.createChildren());
		me.updateLayout();
	},
	updateTitle:function(){
		var me=this;
		var title=me.callParent();
		if(me.jsonData.filterTitle){
			title=title+"&nbsp;-&nbsp;<i>"+me.jsonData.filterTitle+"</i>";
		}
		return title;
	},
	createGrid:function(){
		var me=this;
		var statistic=me.jsonData.statistic;
		var fields=[];
		fields.push({name:'labelV'});
		fields.push({name:'linkV'});
		if(statistic.optionsH){
			for(var i=0;i<statistic.optionsH.length;i++){
				fields.push({name:'label'+i});
				fields.push({name:'link'+i});
			}
		}
		fields.push({name:'total'});

		var columns=[];
		columns.push({
			xtype:'linkcolumn',
			dataIndex: 'labelV',
			handler:me.clickOnLinkV,
			scope:me,
			isLink:me.isLinkV
		});
		if(statistic.optionsH){
			for(var i=0;i<statistic.optionsH.length;i++){
				columns.push({
					xtype:'linkcolumn',
					dataIndex:'label'+i,
					align:'right',
					handler:me.clickOnLink,
					scope:me,
					isLink:me.isLink,
					postProcessRenderer:me.postProcessRendererLink
				});
			}
		}
		columns.push({
			dataIndex: 'total',
			width:50,
			align:'right',
			renderer:me.formatTotal
		});
		columns.push({
			dataIndex: 'dummy',
			flex:1
		});

		var data=[];
		var firstRow={};
		firstRow.labelV="&nbsp;&nbsp;&nbsp;"+statistic.labelH+"</BR>"+statistic.labelV;
		if(statistic.optionsH){
			for(var i=0;i<statistic.optionsH.length;i++){
				firstRow['label'+i]=statistic.optionsH[i].label;
				firstRow['link'+i]=statistic.optionsH[i].id;
			}
		}
		firstRow['total']=getText('twoDimensionalStatistic.total');
		data.push(firstRow);
		if(statistic.optionsV){
			for(var i=0;i<statistic.optionsV.length;i++){
				var row={};
				var option=statistic.optionsV[i];
				row['labelV']=option.label;
				row['linkV']=option.id;
				for(var j=0;j<statistic.optionsH.length;j++){
					var optionValue=statistic.values[i][j];
					row['label'+j]=optionValue.label;
					row['link'+j]=optionValue.id;
				}
				row['total']=statistic.totalV[i];
				data.push(row);
			}
		}
		var lastRow={};
		lastRow.labelV=getText('twoDimensionalStatistic.total');
		if(statistic.totalH){
			for(var i=0;i<statistic.totalH.length;i++){
				lastRow['label'+i]=statistic.totalH[i];
			}
			lastRow['total']=statistic.totalV[statistic.totalV.length-1];
		}
		data.push(lastRow);
		var store = Ext.create('Ext.data.Store', {
			fields: fields,
			data: data
		});

		return Ext.create('Ext.grid.Panel', {
			cls:'dashboardGrid-noBottomBorder',
			store: store,
			columns:columns,
			hideHeaders:true,
			columnLines :false,
			disableSelection: true,
			border:false,
			bodyBorder:false,
			margin:'0 0 0 0',
			viewConfig: {
				stripeRows: true,
				trackOver  :false
			},
			autoHeight:true
		});
	},
	submit:function(fieldValueH,fieldValueV){
		var me=this;
		var urlStr='itemNavigator.action';
		var params={
			'queryType':2,//dashboard
			'queryID':me.jsonData.dashboardID,
			'dashboardParams.dashboardID':me.jsonData.dashboardID
		};
		
		//overwrite projectID and entity type if dashboard is in browse projects page
		if(me.jsonData&&me.jsonData.projectID){
			params['dashboardParams.project']=me.jsonData.projectID;
		}
		if(me.jsonData&&me.jsonData.releaseID){
			params['dashboardParams.release']=me.jsonData.releaseID;
		}
		if(fieldValueH){
			params['dashboardParams.fieldValueH']=fieldValueH;
		}
		if(fieldValueV){
			params['dashboardParams.fieldValueV']=fieldValueV;
		}
		var dummyForm = Ext.create('Ext.form.Panel', {
			items:[],
			url:urlStr,
			standardSubmit:true
		});
		dummyForm.getForm().submit({
			params:params
		});
	},
	clickOnLinkV:function(record,cellIndex){
		var me=this;
		var link=record.data['linkV'];
		me.submit(null,link);
	},
	isLinkV:function(value,metaData,record,rowIndex,colIndex,store,view){
		var link=record.data['linkV'];
		return (link&&link!=="");
	},
	clickOnLink:function(record,cellIndex){
		var me=this;
		var link=record.data['link'+(cellIndex-1)];
		var idx=link.indexOf(";");
		var fieldValueH=null;
		var fieldValueV=null;
		if(idx===-1){
			fieldValueH=link;
		}else{
			fieldValueH=link.substring(0,idx);
			fieldValueV=link.substring(idx+1);
		}
		me.submit(fieldValueH,fieldValueV);
	},
	isLink:function(value,metaData,record,rowIndex,colIndex,store,view){
		var link=record.data['link'+(colIndex-1)];
		if(value==="0"){
			return false;
		}
		return (link&&link!=="");
	},
	postProcessRendererLink:function(value,metaData,record,rowIndex,colIndex,store,view){
		var htmlStr=value;
		var link=record.data['link'+(colIndex-1)];
		if(link&&link!==""){
			if(value==="0"){
				htmlStr='<span style="color:#CACACA;">0</span>';
			}
		}else{
			htmlStr='<B>'+value+'</B>'
		}
		return htmlStr;
	},
	formatTotal:function(value,metaData,record,rowIndex,colIndex,store,view){
		var htmlStr="";
		if(rowIndex>0&&value){
			htmlStr='<B>'+value+'</B>'
		}else{
			htmlStr=value;
		}
		return htmlStr;
	}
});
